File: scripts/installFieldSoilPbr.py
#!/usr/bin/env python3
"""Install FAL PATINA field-soil candidates as runtime PBR state sets.

Raw model output remains immutable in the material-review archive. Runtime
copies correct PATINA's tangent-normal convention, constrain roughness for the
named soil condition, derive conservative cavity AO, and preserve generation
metadata without ever reading or writing the FAL credential.
"""

from __future__ import annotations

import hashlib
import json
import math
import shutil
from dataclasses import dataclass
from pathlib import Path

from PIL import Image, ImageChops, ImageEnhance, ImageFilter, ImageStat


ROOT = Path(__file__).resolve().parents[1]
REVIEW_ROOT = ROOT / "artifacts" / "pbr-material-review" / "patina-candidates"
TARGET_ROOT = ROOT / "public" / "assets" / "textures" / "terrain" / "field_soil_states_v1"


@dataclass(frozen=True)
class FieldSoilMaterial:
    state: str
    slug: str
    normal_strength: float
    roughness_minimum: float
    roughness_maximum: float
    metres_per_tile: float


MATERIALS = (
    FieldSoilMaterial("ploughed", "field-soil-fresh-ploughed-v1", 0.58, 0.74, 0.96, 2.8),
    FieldSoilMaterial("seedbed", "field-soil-fine-seedbed-v1", 0.40, 0.84, 0.99, 2.4),
    FieldSoilMaterial("fallow", "field-soil-weathered-fallow-v1", 0.34, 0.82, 0.99, 3.2),
    FieldSoilMaterial("harvested", "field-soil-dry-harvested-v3", 0.32, 0.86, 1.00, 2.9),
)


def require(path: Path) -> Path:
    if not path.is_file():
        raise FileNotFoundError(path)
    return path


def copy(source: Path, target: Path) -> None:
    target.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(require(source), target)


def wrap_blur(image: Image.Image, radius: float) -> Image.Image:
    image = image.convert("L")
    width, height = image.size
    tiled = Image.new("L", (width * 3, height * 3))
    for y in range(3):
        for x in range(3):
            tiled.paste(image, (x * width, y * height))
    blurred = tiled.filter(ImageFilter.GaussianBlur(radius))
    return blurred.crop((width, height, width * 2, height * 2))


def derive_ao(height_path: Path, output_path: Path) -> None:
    height = Image.open(require(height_path)).convert("L")
    cavities = []
    for radius, gain in ((3.0, 2.0), (12.0, 1.55), (36.0, 1.15)):
        neighbourhood = wrap_blur(height, radius)
        cavity = ImageChops.subtract(neighbourhood, height)
        cavity = cavity.point(lambda value, scale=gain: min(255, round(value * scale)))
        cavities.append(cavity)
    combined = ImageChops.lighter(cavities[0], cavities[1])
    combined = ImageChops.lighter(combined, cavities[2])
    combined = ImageEnhance.Contrast(combined).enhance(1.22)
    combined.point(lambda value: max(154, 255 - round(value * 0.72))).save(output_path)


def prepare_runtime_normal(source: Path, target: Path, strength: float) -> dict[str, object]:
    image = Image.open(require(source)).convert("RGB")
    source_mean = ImageStat.Stat(image).mean
    mean_x = source_mean[0] / 127.5 - 1.0
    mean_y = -(source_mean[1] / 127.5 - 1.0)
    output = bytearray(image.width * image.height * 3)
    offset = 0
    pixels = image.get_flattened_data() if hasattr(image, "get_flattened_data") else image.getdata()
    for red, green, _blue in pixels:
        nx = (red / 127.5 - 1.0 - mean_x) * strength
        ny = (-(green / 127.5 - 1.0) - mean_y) * strength
        inverse_length = 1.0 / math.sqrt(nx * nx + ny * ny + 1.0)
        output[offset] = round((nx * inverse_length * 0.5 + 0.5) * 255)
        output[offset + 1] = round((ny * inverse_length * 0.5 + 0.5) * 255)
        output[offset + 2] = round((inverse_length * 0.5 + 0.5) * 255)
        offset += 3
    runtime = Image.frombytes("RGB", image.size, bytes(output))
    runtime.save(target, optimize=True)
    return {
        "greenChannelFlipped": True,
        "meanTiltRemoved": True,
        "xyStrength": strength,
        "sourceMeanRgb": [round(value, 3) for value in source_mean],
        "runtimeMeanRgb": [round(value, 3) for value in ImageStat.Stat(runtime).mean],
    }


def prepare_roughness(
    source: Path,
    target: Path,
    minimum: float,
    maximum: float,
) -> None:
    image = Image.open(require(source)).convert("L")
    image.point(
        lambda value: round((minimum + (maximum - minimum) * (value / 255)) * 255)
    ).save(target, optimize=True)


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        while block := handle.read(1024 * 1024):
            digest.update(block)
    return digest.hexdigest()


def install(material: FieldSoilMaterial) -> dict[str, object]:
    source = REVIEW_ROOT / material.slug
    target = TARGET_ROOT / material.state
    target.mkdir(parents=True, exist_ok=True)
    copy(source / "basecolor.png", target / "albedo.png")
    normal_processing = prepare_runtime_normal(
        source / "normal.png",
        target / "normal.png",
        material.normal_strength,
    )
    prepare_roughness(
        source / "roughness.png",
        target / "roughness.png",
        material.roughness_minimum,
        material.roughness_maximum,
    )
    copy(source / "height.png", target / "height.png")
    derive_ao(source / "height.png", target / "ao.png")
    Image.new("L", Image.open(source / "basecolor.png").size, 0).save(
        target / "metalness.png",
        optimize=True,
    )
    copy(source / "generation.json", target / "generation.json")
    files = {
        name: sha256(target / name)
        for name in ("albedo.png", "normal.png", "roughness.png", "ao.png", "height.png", "metalness.png")
    }
    metadata = {
        "state": material.state,
        "sourceModel": "fal-ai/patina/material",
        "reviewCandidate": material.slug,
        "metresPerTile": material.metres_per_tile,
        "normalProcessing": normal_processing,
        "roughnessRange": [material.roughness_minimum, material.roughness_maximum],
        "metalness": 0,
        "maps": files,
    }
    (target / "runtime-material.json").write_text(
        json.dumps(metadata, indent=2) + "\n",
        encoding="utf8",
    )
    (target / "README.md").write_text(
        f"""# {material.state.title()} field soil PBR

- Generated by `fal-ai/patina/material` from `{material.slug}`.
- Physical texture scale: `{material.metres_per_tile:.1f}` metres per tile.
- Base color is sRGB; normal, roughness, AO, height, and metalness are linear.
- Runtime normal flips PATINA's green channel, removes mean tilt, and limits XY relief to `{material.normal_strength:.2f}`.
- Roughness is constrained to `{material.roughness_minimum:.2f}–{material.roughness_maximum:.2f}`; metalness is zero.
- Raw API output and its exact prompt remain in `artifacts/pbr-material-review/patina-candidates/{material.slug}/`.
""",
        encoding="utf8",
    )
    return metadata


def main() -> None:
    TARGET_ROOT.mkdir(parents=True, exist_ok=True)
    installed = [install(material) for material in MATERIALS]
    manifest = {
        "version": 1,
        "coordinateDomain": "world-xz-metres",
        "antiTiling": "shared low-frequency coordinate warp across all PBR channels",
        "edgeTransition": "continuous fieldEdgeBlend alpha over a deterministic irregular metre-scale band",
        "states": {entry["state"]: entry for entry in installed},
        "growing": {
            "source": "../mammoth_terrain_dirt",
            "reason": "Established crop soil reuses the approved backyard garden-bed loam.",
            "metresPerTile": 2.2,
        },
        "unploughed": {
            "source": "native terrain",
            "reason": "Unworked land does not receive an artificial cultivated-soil overlay.",
        },
    }
    (TARGET_ROOT / "manifest.json").write_text(
        json.dumps(manifest, indent=2) + "\n",
        encoding="utf8",
    )
    print(f"Installed {len(installed)} field-soil PBR state sets in {TARGET_ROOT}")


if __name__ == "__main__":
    main()

File: scripts/installFieldSoilPbr.mjs
#!/usr/bin/env node
/** Install FAL PATINA field-soil candidates as runtime PBR state sets. */

import { createHash } from 'node:crypto';
import { execFile } from 'node:child_process';
import { copyFile, mkdir, mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join, resolve } from 'node:path';
import { promisify } from 'node:util';

const execFileAsync = promisify(execFile);
const ROOT = resolve(import.meta.dirname, '..');
const REVIEW_ROOT = join(ROOT, 'artifacts', 'pbr-material-review', 'patina-candidates');
const TARGET_ROOT = join(ROOT, 'public', 'assets', 'textures', 'terrain', 'field_soil_states_v1');

const materials = [
  { state: 'ploughed', slug: 'field-soil-fresh-ploughed-v1', normalStrength: 0.58, roughness: [0.74, 0.96], metresPerTile: 2.8 },
  { state: 'seedbed', slug: 'field-soil-fine-seedbed-v1', normalStrength: 0.4, roughness: [0.84, 0.99], metresPerTile: 2.4 },
  { state: 'fallow', slug: 'field-soil-weathered-fallow-v1', normalStrength: 0.34, roughness: [0.82, 0.99], metresPerTile: 3.2 },
  { state: 'harvested', slug: 'field-soil-dry-harvested-v3', normalStrength: 0.32, roughness: [0.86, 1], metresPerTile: 2.9 },
];

async function magick(...args) {
  await execFileAsync('magick', args, { windowsHide: true, maxBuffer: 8 * 1024 * 1024 });
}

async function channelMean(path, channel) {
  const { stdout } = await execFileAsync(
    'magick',
    [path, '-alpha', 'off', '-format', `%[fx:mean.${channel}]`, 'info:'],
    { windowsHide: true },
  );
  return Number.parseFloat(stdout.trim());
}

async function prepareRuntimeNormal(source, target, strength, temporaryRoot) {
  const meanR = await channelMean(source, 'r');
  const meanG = await channelMean(source, 'g');
  const red = join(temporaryRoot, 'normal-r.png');
  const green = join(temporaryRoot, 'normal-g.png');
  const blue = join(temporaryRoot, 'normal-b.png');
  await magick(source, '-alpha', 'off', '-channel', 'R', '-separate',
    '-fx', `clamp(0.5 + (u - ${meanR}) * ${strength})`, red);
  await magick(source, '-alpha', 'off', '-channel', 'G', '-separate',
    '-fx', `clamp(0.5 + ((${meanG}) - u) * ${strength})`, green);
  await magick(source, '-alpha', 'off', '-channel', 'B', '-separate',
    '-fill', 'white', '-colorize', '100', blue);
  await magick(red, green, blue, '-combine', '-define', 'png:color-type=2', target);
  return {
    greenChannelFlipped: true,
    meanTiltRemoved: true,
    xyStrength: strength,
    sourceMeanRG: [Number(meanR.toFixed(6)), Number(meanG.toFixed(6))],
  };
}

async function prepareRoughness(source, target, minimum, maximum) {
  await magick(
    source, '-alpha', 'off', '-colorspace', 'Gray',
    '-fx', `${minimum}+(${maximum - minimum})*u`,
    '-define', 'png:color-type=0', target,
  );
}

async function deriveAo(source, target) {
  // PATINA height owns the same clods as the other maps. This conservative
  // remap keeps cavities darker without introducing an unrelated noise field.
  await magick(
    source, '-alpha', 'off', '-colorspace', 'Gray',
    '-fx', '0.68+0.32*u', '-define', 'png:color-type=0', target,
  );
}

async function prepareMetalness(source, target) {
  await magick(
    source, '-alpha', 'off', '-colorspace', 'Gray',
    '-fill', 'black', '-colorize', '100',
    '-define', 'png:color-type=0', target,
  );
}

async function sha256(path) {
  return createHash('sha256').update(await readFile(path)).digest('hex');
}

async function install(material, temporaryRoot) {
  const source = join(REVIEW_ROOT, material.slug);
  const target = join(TARGET_ROOT, material.state);
  await mkdir(target, { recursive: true });
  await copyFile(join(source, 'basecolor.png'), join(target, 'albedo.png'));
  const normalProcessing = await prepareRuntimeNormal(
    join(source, 'normal.png'), join(target, 'normal.png'),
    material.normalStrength, temporaryRoot,
  );
  await prepareRoughness(
    join(source, 'roughness.png'), join(target, 'roughness.png'),
    material.roughness[0], material.roughness[1],
  );
  await copyFile(join(source, 'height.png'), join(target, 'height.png'));
  await deriveAo(join(source, 'height.png'), join(target, 'ao.png'));
  await prepareMetalness(join(source, 'basecolor.png'), join(target, 'metalness.png'));
  await copyFile(join(source, 'generation.json'), join(target, 'generation.json'));
  const fileNames = ['albedo.png', 'normal.png', 'roughness.png', 'ao.png', 'height.png', 'metalness.png'];
  const files = Object.fromEntries(await Promise.all(fileNames.map(async (name) => [
    name, await sha256(join(target, name)),
  ])));
  const metadata = {
    state: material.state,
    sourceModel: 'fal-ai/patina/material',
    reviewCandidate: material.slug,
    metresPerTile: material.metresPerTile,
    normalProcessing,
    roughnessRange: material.roughness,
    metalness: 0,
    maps: files,
  };
  await writeFile(join(target, 'runtime-material.json'), `${JSON.stringify(metadata, null, 2)}\n`);
  await writeFile(join(target, 'README.md'), `# ${material.state[0].toUpperCase()}${material.state.slice(1)} field soil PBR

- Generated by \`fal-ai/patina/material\` from \`${material.slug}\`.
- Physical texture scale: \`${material.metresPerTile.toFixed(1)}\` metres per tile.
- Base color is sRGB; normal, roughness, AO, height, and metalness are linear.
- Runtime normal flips PATINA's green channel, removes mean tilt, and limits XY relief to \`${material.normalStrength.toFixed(2)}\`.
- Roughness is constrained to \`${material.roughness[0].toFixed(2)}–${material.roughness[1].toFixed(2)}\`; metalness is zero.
- Raw API output and its exact prompt remain in \`artifacts/pbr-material-review/patina-candidates/${material.slug}/\`.
`);
  return metadata;
}

async function main() {
  await mkdir(TARGET_ROOT, { recursive: true });
  const temporaryRoot = await mkdtemp(join(tmpdir(), 'field-soil-pbr-'));
  try {
    const installed = [];
    for (const material of materials) installed.push(await install(material, temporaryRoot));
    const manifest = {
      version: 1,
      coordinateDomain: 'world-xz-metres',
      antiTiling: 'shared low-frequency coordinate warp across all PBR channels',
      edgeTransition: 'continuous fieldEdgeBlend alpha over a deterministic irregular metre-scale band',
      states: Object.fromEntries(installed.map((entry) => [entry.state, entry])),
      growing: {
        source: '../mammoth_terrain_dirt',
        reason: 'Established crop soil reuses the approved backyard garden-bed loam.',
        metresPerTile: 2.2,
      },
      unploughed: {
        source: 'native terrain',
        reason: 'Unworked land does not receive an artificial cultivated-soil overlay.',
      },
    };
    await writeFile(join(TARGET_ROOT, 'manifest.json'), `${JSON.stringify(manifest, null, 2)}\n`);
    process.stdout.write(`Installed ${installed.length} field-soil PBR state sets in ${TARGET_ROOT}\n`);
  } finally {
    await rm(temporaryRoot, { recursive: true, force: true });
  }
}

await main();
